### PR TITLE
feat(skills): sync project-board Status from /gh-pr and /gh-commit

### LINKED ISSUE #199: [Feature] /gh-pr·/gh-commit 에 projectV2 Status 자동 sync — 공용 헬퍼 추출 및 가드 옵션 추가
## Context

칸반 보드 (GitHub Projects v2 #3, `dEitY719/dotfiles`) 의 Status 자동
동기화가 현재는 **`/gh-flow` 경로에서만** 작동한다. 분리 실행 (`/gh-commit`
+ `/gh-pr`) 시엔 sync 가 발화하지 않아 카드가 `Backlog` 에 고정되는
문제가 있다.

본 세션에서 발견된 사례: `/gh-commit` → `/gh-pr` 로 생성한 PR #197 이
`Auto-add to project` 덕분에 보드에는 올라갔지만, `Backlog → In review`
전환이 수동 상태로 남아 보이지 않는 것처럼 느껴졌다.

### 실제 현 상황

- `shell-common/functions/gh_flow.sh:522` 에 정의된 내부 헬퍼
  `_gh_flow_set_project_status` 가 gh-flow worker 두 지점
  (implement 시작 / PR 번호 확보 직후) 에서만 호출된다.
- `claude/skills/gh-pr/SKILL.md` · `claude/skills/gh-commit/SKILL.md`
  에는 해당 호출이 없다.
- SSOT (`docs/standards/github-project-board.md`) 와 playbook
  (`docs/playbooks/kanban-board-setup.md`) 은 두 전환을 모두
  "**수동**" 으로 서술 — gh-flow 자동화 도입 (커밋 `fc48e7a`)
  이후 반영되지 않은 상태다.

### 요구사항

- `/gh-pr` 단독 실행 시에도 PR 카드가 `Backlog → In review` 로 자동 전환.
- `/gh-commit` 단독 실행 시 해당 커밋이 최초로 이슈에 연결되는 시점엔
  Issue 카드가 `Backlog → In progress` 로 자동 전환 (단, 역행 방지
  가드 필수 — `/gh-commit` 은 동일 이슈에 여러 번 호출됨).
- **`/gh-pr` 은 `dotfiles` 외 다른 프로젝트 (예: 프로젝트 A, 보드 없음)
  에서도 사용 중** — 보드가 없는 레포에선 자동 감지로 no-op 이어야
  한다. 추가 분기 없이 기존 헬퍼 설계가 이미 이를 만족한다
  (`projectItems` 가 0건이면 조용히 return 0).

### 1. 헬퍼 리팩터 — `_gh_flow_set_project_status` 를 공용 함수로 추출

**이동**: `shell-common/functions/gh_flow.sh` → `shell-common/functions/gh_project_status.sh` (신규 파일)

**개명**: `_gh_flow_set_project_status` → `_gh_project_status_sync` (gh-flow 한정 함수 같은 오해 제거)

**환경 변수**: `GH_FLOW_PROJECT_STATUS_SYNC` → `GH_PROJECT_STATUS_SYNC` (기존 변수명도 하위호환 유지: 둘 중 하나가 `0` 이면 opt-out)

**stderr prefix**: `[gh-flow-worker]` → `[gh-project-status]`

**새 옵션**: `--only-from <Status>[,<Status>...]` — 현재 Status 가 지정 리스트에 포함된 경우만 천이. 역행 방지.

**업데이트된 시그니처**:
```
_gh_project_status_sync <issue|pr> <number> <target-status> [--only-from <status-list>]
```

`gh_flow.sh` 기존 두 호출점 (lines 412, 454) 을 새 이름으로 전환. 기존 semantic 은 옵션 미사용 상태로 보존 (gh-flow worker 는 항상 실행).

### 2. `/gh-pr` 스킬 확장

`claude/skills/gh-pr/SKILL.md` Step 7 (PR URL 리턴) **직전**에 호출:

```
_gh_project_status_sync pr $PR_NUM "In review"
```

- 가드 없음. PR 라이프사이클은 단선이라 `/gh-pr` 시점에 idempotent 하게 `In review` 고정.
- 보드 없는 repo (프로젝트 A 등) → 헬퍼가 조용히 no-op.

### 3. `/gh-commit` 스킬 확장

`claude/skills/gh-commit/SKILL.md` 커밋 성공 후, commit message 에서 resolve 된 issue 번호가 있으면:

```
_gh_project_status_sync issue $N "In progress" --only-from Backlog
```

- **`--only-from Backlog` 가드 필수**: `/gh-commit` 은 동일 이슈 브랜치에 여러 번 호출된다. 이슈가 이미 `In review` 로 넘어간 뒤의 fix 커밋에서 `In progress` 로 역행시키면 안 됨.
- commit message 에서 `Closes|Fixes|Refs #N` 가 없으면 sync 스킵.

### 4. 문서 정정

- `docs/playbooks/kanban-board-setup.md`
  - §5 라이프사이클 다이어그램: `[수동]` 라벨 제거 및 재작성.
  - §6 자동 vs 수동: "수동" 항목 3개를 "`/gh-flow`·`/gh-pr`·`/gh-commit` 자동 / raw git 사용 시 수동" 로 교체.
- `docs/standards/github-project-board.md`
  - §수동 이동: 제목/본문 "자동 전환 규칙 (스킬 경유)" 로 리프레이밍.
  - §운영 상의 유의사항 lines 210-217: 동일 기조로 업데이트.
- 공통: "보드 미연결 repo (예: 다른 프로젝트) 에서는 헬퍼가 자동으로 no-op" 각주 추가.

### 5. 테스트

- `_gh_project_status_sync` 에 대한 bats 스모크 테스트 확장:
  - 기존 케이스 (opt-out, invalid kind, missing args) 유지
  - `--only-from` 옵션 파싱 테스트
  - `--only-from` 가드 발화/통과 시나리오
- Integration 테스트는 생략 (실제 프로젝트 mutation 은 fixture 로 대체 어려움)

### 세션에서 확인된 사실

- PR #197 이 보드에 안 보인다고 느꼈지만 실제론 `Backlog` 에 있음 —
  필터/그룹 설정이나 컬럼 접힘 때문에 가시성만 떨어진 상태였음.
- 원인은 `/gh-commit` + `/gh-pr` 분리 실행 시 `_gh_flow_set_project_status`
  가 호출되지 않는 것. gh-flow worker 경로에서만 sync 가 일어난다.
- 기존 헬퍼는 이미 auto-discovery 덕에 "보드 없는 repo" 를 자동
  처리한다 (`projectItems(first: 10)` 0건 → `_triples` 빈 문자열 →
  조용히 return 0).

### 대안 검토

**Option A — 문서만 정정**: "gh-flow 쓰면 자동, 안 쓰면 수동" 로 표기.
**단점**: 실행 경로마다 보드 상태가 달라져 운영 혼란. 거절.

**Option B — 스킬 확장 + 문서 정정** (채택): 경로 무관하게 일관.
하위 선택:
- B1: `/gh-pr` 만 확장. **단점**: `/gh-commit` 단독 실행 시 Issue In progress
  수동 남음.
- B2: 둘 다 확장 + `/gh-commit` 은 `--only-from Backlog` 가드. **채택**.

### 네이밍 결정

- 함수명: `_gh_project_status_sync` — `_gh_flow_` prefix 는 caller 오인.
- 옵션명: `--only-from` — `--guard-from`, `--require-current` 등 검토했으나
  읽었을 때 의도가 가장 분명한 `--only-from` 채택.
- 파일 경로: `shell-common/functions/gh_project_status.sh` — `gh_flow.sh`
  는 worker 자체에 집중, 재사용 헬퍼는 분리.

## Open Questions

1. **하위 호환 기간**: `_gh_flow_set_project_status` 를 즉시 제거할지,
   deprecation wrapper 로 남길지. dotfiles 는 solo repo 라 즉시 제거가
   일반적이지만, 외부에서 헬퍼를 import 했다면 깨진다. → 즉시 제거안이
   기본, 대안 제시 가능.

2. **`GH_FLOW_PROJECT_STATUS_SYNC` 하위호환**: 당장은 유지하되 "deprecated,
   use `GH_PROJECT_STATUS_SYNC`" 경고를 stderr 에 낼지 여부.

3. **`/gh-pr-reply` 에도 sync 필요한가**: Changes requested 루프
   (`In progress ↔ In review`) 는 현재 Projects v2 빌트인 `Code changes
   requested` 워크플로우가 다루므로 신규 sync 불필요 — 검증만 필요.

4. **`/gh-issue-create` 도 확장?** 새 이슈를 보드에 추가 직후 상태를
   `Backlog` 이 아닌 다른 값으로 초기화 하는 use-case 는 없어 보임 —
   이번 scope 제외.

## References

- 관련 파일:
  - `shell-common/functions/gh_flow.sh:522` (이동 대상 헬퍼)
  - `shell-common/functions/gh_flow.sh:412,454` (업데이트 대상 호출점)
  - `claude/skills/gh-pr/SKILL.md` (Step 7 확장)
  - `claude/skills/gh-commit/SKILL.md` (Step 5 확장)
  - `docs/playbooks/kanban-board-setup.md` (§5, §6 정정)
  - `docs/standards/github-project-board.md` (§수동 이동, §운영 유의사항 정정)
- 관련 커밋:
  - `fc48e7a feat(gh-flow): sync project board Status on worker transitions` — 현 헬퍼 도입.
- 관련 PR: #197 (현재 오픈, Auto-archive items 문서화 — 본 이슈와 독립).
- 트리거 세션: PR #197 생성 후 Backlog 에 머문 현상 관찰 대화
  (2026-04-24).

---

📊 ~1500 tokens · 👤 ~2 h · 🤖 ~6 min

### PR DESCRIPTION
**Pull request by dEitY719**
## Summary
- Extract `_gh_flow_set_project_status` from `gh_flow.sh` into a shared helper `_gh_project_status_sync` (`shell-common/functions/gh_project_status.sh`) so `/gh-pr` and `/gh-commit` (the atomic skills, not just the `/gh-flow` worker) can also push projectV2 board transitions.
- Add an `--only-from <list>` guard so `/gh-commit`'s `Backlog → In progress` transition only fires from `Backlog`. Without it, follow-up fix commits to an open PR would bounce the issue card from `In review` back to `In progress` each time.
- Repos with no projectV2 attached are auto-detected (helper finds zero project items, returns 0 silently) — no behavioral impact on side projects without a board.

## Changes
- `shell-common/functions/gh_project_status.sh` (new): shared helper with arg parsing, opt-out (`GH_PROJECT_STATUS_SYNC=0`, plus legacy `GH_FLOW_PROJECT_STATUS_SYNC=0` for back-compat), single GraphQL query that also returns the current Status (needed for `--only-from`), and the `_gh_project_status_in_list` membership predicate (preserves Status names with internal spaces like `In progress`).
- `shell-common/functions/gh_flow.sh`: drop the inline `_gh_flow_set_project_status` body; worker now calls `_gh_project_status_sync` for both Step 2a (`Issue → In progress`) and post-Step 2c (`PR → In review`).
- `claude/skills/gh-pr/SKILL.md`: insert new Step 7 (`PR → In review` sync) and renumber Report to Step 8.
- `claude/skills/gh-commit/SKILL.md`: insert new Step 5 (`Issue Backlog → In progress` with mandatory `--only-from Backlog` guard) and renumber Verify to Step 6.
- `docs/standards/github-project-board.md`, `docs/playbooks/kanban-board-setup.md`: update the automation matrix and "수동 vs 자동" tables to reflect that the two entry transitions are now skill-driven (with no-op fallback on board-less repos).
- `tests/bats/functions/gh_project_status.bats` (new, 13 cases): loading in bash + zsh, opt-out via both env vars, arg validation, `--only-from` parsing edge cases, and `_gh_project_status_in_list` membership semantics including the `In progress` whitespace regression.
- `tests/bats/functions/gh_flow.bats`: drop the now-orphaned `_gh_flow_set_project_status` cases (coverage moved with the helper).

## Test plan
- [x] `bats tests/bats/functions/gh_project_status.bats` → 13/13 pass
- [x] `bats tests/bats/functions/gh_flow.bats` → still passes after the helper extraction
- [x] `bats tests/bats/functions/` full suite green
- [x] `shellcheck` + `bash -n` + `zsh -n` clean on the new helper
- [ ] Live: run `/gh-pr 199` against this branch and confirm the PR card moves to `In review` on the project board (this very PR is the smoke test)
- [ ] Live: run `/gh-commit` on a follow-up tweak to issue #199 and confirm `--only-from Backlog` guard skips the transition (issue stays at `In review`, no regression to `In progress`)
- [ ] Live: invoke `/gh-pr` or `/gh-commit` in a side repo without a projectV2 board → expect silent no-op

## Related
Closes #199

---

📊 ~1000 tokens · 👤 ~8 h · 🤖 ~24 min

### CODE REVIEW

File: shell-common/functions/gh_project_status.sh [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                    }
+                  }
+                }
+              }
+            }
+          }" \
+        -f owner="$_owner" -f repo="$_repo" -F number="$_num" -f target="$_target" \
+        --jq ".data.repository.${_q_field}.projectItems.nodes[]
+              | select(.project.field.options? | length > 0)
+              | \"\(.project.id)|\(.id)|\(.project.field.id)|\(.project.field.options[0].id)|\(.fieldValueByName.name // \"\")\"" \
```
[the remark this answers is not in the document]
Comment (dEitY719) [reply shell-common/functions/gh_project_status.sh]: Declined. The preceding `select(.project.field.options? | length > 0)` already filters out any node where `.project`, `.project.field`, or `.options` is null — the trailing `?` suppresses the error and `null | length` evaluates to 0, so `select` rejects the row. After the filter passes, `.project.id`, `.project.field.id`, and `.project.field.options[0].id` are guaranteed non-null. The remaining `.fieldValueByName.name // ""` is also safe because jq's basic field access returns `null` on `null` rather than erroring, and `// ""` defaults it. Adding `?` on every path component would be defensive noise without changing behavior.

[the remark this answers is not in the document]
Comment (dEitY719) [reply shell-common/functions/gh_project_status.sh]: Accepted. Fixed in ddecc7f — added jq optional chaining (`?`) on `.project`, `.field`, `.options`, `.options[0]`, and `.fieldValueByName` accesses in the projectItems extraction filter, matching the suggestion verbatim. The upstream `select(... | length > 0)` was already filtering null cases, but the explicit `?` operators make the path-traversal robust against edge cases where an intermediate nullable field returns an unexpected type.
